### ISSUE TITLE
[wishlist] parametric tests

### ISSUE THREAD

**Issue by yarikoptic** [bats-core/bats-core#241]
It is quite often desired to run the same test logic in various scenarios.  ATM there is no dedicated facility in bats to make that easy, so [I resorted to placing test logic into a dedicated function and ugly copy/pasting for invocations with different parameters where I needed it](https://github.com/ReproNim/containers/pull/27/files#diff-308e0083e635922e293be5cadbc1000bR46)

I am not yet sure what could be the "bats way" to specify parametric tests but  since there is parsing and preparation of the test code is happening anyways could be something like

```shell
BACKEND=( docker singularity )
NJOBS=( 1 10 )

@test "description" BACKEND NJOBS {
   echo "Here I am with backend=$backend njobs=$njobs"
   ...
}
```

and running it would result in 4 tests with adjusted descriptions:

- `description [backend=docker,njobs=1]`
- `description [backend=docker,njobs=10]`
- `description [backend=singularity,njobs=1]`
- `description [backend=singularity,njobs=10]`

and still subject to `--filter` selection (so e.g. I could select only the ones for `backend=docker`)

**Comment by sdolenc** [comment]
I would also like to see this feature added 👍 . In the meantime, you could use a wrapper test which provides parameters (as environment variables) to the actual test. Here's how I would implement your example

wrapper.bats

```
#!/usr/bin/env bats

cd $BATS_TEST_DIRNAME

@test "wrapper" {
  final_status=0

  BACKENDS=( "docker" "singularity" )
  NJOBS=( 1 10 )

  for b in "${BACKENDS[@]}"; do
    for nj in "${NJOBS[@]}"; do
      backend="$b" njobs=$nj run bats -t test.bats

      echo "# $output" >&3
      echo "#" >&3
      final_status+=$status
    done
  done

  [ "$final_status" -eq 0 ]
}
```

test.bats

```
#!/usr/bin/env bats

@test "description [backend=$backend,njobs=$njobs]" {
  echo "# Here I am with backend=$backend njobs=$njobs" >&3
}
```
running  `bats -t wrapper.bats` results in

```
1..1
# 1..1
# Here I am with backend=docker njobs=1
ok 1 description [backend=docker,njobs=1]
#
# 1..1
# Here I am with backend=docker njobs=10
ok 1 description [backend=docker,njobs=10]
#
# 1..1
# Here I am with backend=singularity njobs=1
ok 1 description [backend=singularity,njobs=1]
#
# 1..1
# Here I am with backend=singularity njobs=10
ok 1 description [backend=singularity,njobs=10]
#
ok 1 wrapper
```
I hope this helps

**Comment by dimo414** [comment]
This would be appreciated by me too. I'd be happy to draft a PR if there's a design we can agree on.

The proposed `@test "description" BACKEND NJOBS {` syntax backed by arrays seems like a good approach to me.

One thing I'd change is the different-casing of the parameter vs. the injected argument. I'd suggest shadowing the same name (i.e. `BACKEND` instead of `backend`) with a `local` inside the generated function, so that there's no need to transform strings or accidentally access the array\*.

---

\* with the initial proposal, a user could easily mistakenly write `echo "Here I am with backend=$BACKEND njobs=$NJOBS"` which would unconditionally access the 0-th index of the global, instead of the parameter value set by the test.

**Comment by sdolenc** [comment]
Sounds good to me! @dimo414 , I'd be happy to test your draft while it's in development. Let me know if there are other ways I could help

**Comment by dimo414** [comment]
I just want to make sure there's buy-in from core, and flesh out any further design constraints.

**Comment by dimo414** [comment]
I've chatted with @sublimino and we're talking about a couple of approaches (which I'll share in more detail later), but one issue that needs to be settled before I can move forward is what *type(s)* of parameterization are we looking to support?

* **Cartesian** the proposal above (`BACKEND NJOBS`) would test all combinations of the given variables, similar to GitHub Action's [matrix](https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategy)
* **Sequential** each set of parameters to test is specified separately, and each denotes a single test; this is more typical, e.g. [JUnit](https://github.com/junit-team/junit4/wiki/Parameterized-tests), [googletest](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests), and [pytest](https://docs.pytest.org/en/latest/parametrize.html#parametrize-basics)

Can folks share their use-cases for parameterized testing? I would expect that sequential testing is more commonly needed (notably to compare a sequence of input:expected-output pairs), but it would be helpful to understand what people need. It's totally possible to support both if there's interest!

**Comment by sdolenc** [comment]
My recent use case for parameterized tests in bats was Sequential. I implemented tests using a variation of the workaround I mentioned above. The most relevant line of code:
https://github.com/kdabir/has/blob/4ccd0177d458800f7cc721df08c2f0ad643990ac/tests/test_all_packages.bats#L25
And the entire change: https://github.com/kdabir/has/pull/43

I can see the merit for Cartesian parameters (supporting an entire matrix of test inputs sounds awesome), but I have personally only needed/written Sequential parameterized tests.

**Comment by martin-schulze-vireso** [comment]
Some test frameworks use a data generation function that creates the list of parameters to test. Having a function would allow for sequential and matrix data. However, it might be more complicated to implement the parameter passing.

**Comment by dimo414** [comment]
Apologies for the radio silence, I'm still very much interested in addressing this use case but haven't had time to properly tackle it.

> Why not simply rely on the numbered function parameters

That's definitely an option; I was mostly concerned with the `BACKEND` -> `backend` transformation which seems error-prone. The benefit of shadowing is clearer names (not having to remember what `$3` is supposed to mean) but I agree it may be the most straightforward solution. Especially if we build parameterization on top of dynamic registration (#306) which has been the direction I've been favoring it's likely we'd use positional arguments.

**Comment by martin-schulze-vireso** [comment]
@dimo414 You could have a look at #349 it is still very rough but I believe it offers enough flexibilty to achieve what you want. I think it could be combined with a helper that does the looping over the given list of arrays' values.

**Comment by martin-schulze-vireso** [comment]
This seems to be a duplicate of #135

**Comment by irkode** [comment]
Hi all,
Seen the issue is open a long time and it was very quiet here. Also no branch seen.
Didn't know how to pop in the discussion soSo I though I start playing with a POC implementation
will improve my bash and bats skills ;-)

If you think it may be something valuable i would be please to continue, participate...
just tested on `GNU bash, version 5.1.16(1)-release` and I dunno if the code is real good.
The bats test suite passes incl my added tests.

```shell
    # initialize list of parameters to use
    setup() {
        declare -ga Values=('p1' 'p2')
    }

    @test "T0" {
      # this will just execute as is
    }

    # next test cases will be turned into parameterized tests with params from array Values
    # bats param_set list 2 Values
    # will be executed two times one with 'p1' and one with 'p2'
    @test "T1" {
        # this will be the parameter from Values. choose a name you like
        param=${1-}
    }

    # param_set is still active, so this will also produce two test runs
    @test "T2" {
        param=${1-}
    }

    # bats param_end
    # turn off parameterization and continue

    @test "T3" {
      # this will just execute as is
    }
```
resulting in output like this:
```
A test runner will be created for each parameter entry which at execution time calls
the defined test function passing one of the array values.
So for TAP output you will get something like this.

    1..6
    ok 1 T0
    ok 2 T0 run 0
    ok 3 T0 run 1
    ok 4 T2 run 0
    ok 5 T2 run 1
    ok 6 T3
```
It is of now is able to generate parameterized tests based on list values. The design should allow to add 2D or Matrix tests without big changes.

Bats is the heck of a bash script and I did not want to dive to deep into all the evaluation, trap, logic. So i thought about just injecting  tests based on some template...

Please have a look at https://github.com/irkode/bats-core/tree/parameterized-tests to see it

The implementation has two main parts
* change bats-preprocess to accept two new _bats command comments_ (`param_set` and `param_end`)
  * based on that - and an array definition in test `setup()` file  - it will generate test runners for each list value
* change stack trace printing to support the runners and parameterized test cases

I added a short text to `man/bats.7.ronn` and also added tests for the implementation. to `test/params.bats` and `test/fixtures/params` . Guess these will help to see what it does.

**Comment by martin-schulze-vireso** [comment]
@irkode I am sorry for the long wait on this. It's been on the short list for a long time but always was pushed back. I had a quick look at your code and have following observations:

1. I'd expect the values to be declared closer to the test. Using `setup` seems counterintuitive to me, because that runs after the "default" test registration
2. duplicating the amount of parameters is not user-friendly and at worst leads to missed parameters
3. It would be nice to see the pararmeter values in the TAP output (but that might become unwieldy)

Finally, I want to refer to #306 which already has a (stale) PR. I think both of these wishes might interact. I always saw that issue as a good foundation for implementing this one and as a stop gap solution to enable users to set up their own parametrization.

**Comment by irkode** [comment]
Hi @martin-schulze-vireso,

Thx for checking it out. I got your points.

if I got it right, #306 and #349 use eval and dynamic test generation. Due to their definition, the actual test function is not declared using '@test' and looping is done by the user.

So kind o different approach to solve the puzzle.

The implementation intention of the PoC was to do

* minimize modification to the `bats-core` code as possible (`bats-preprocess` only currently, all other changes just formatting)
* no code evaluation in `bats-preprocess`
* keep the _line by line_ parsing logic
* easy possibility to define parameters
* provide parameterized tests with a style `known by users`.
    * define parameters
    * declare a @test as parameterized

1.  `setup()` is executed _before_ each test, so it looked like the ideal place.
    I'm no sure what you mean with "default" test registration, but there might be better places to hook in i don't be aware of.

    Seperating  _declare,  define and use_  is common, so did not I vene thought about before, but I still think this should not be a problem in general.  *But* you're right: the implementation automatically keeps `param_set`  which will lead to confusion in files with multiple parameter arrays or weird mixing of `param_set`, `param_end` and `@test`

2. not evaluating the parameters while `bats_processing` will shift all checks to the test execution. There's already a check in the runner for
    * undefined array
    * undefined elements.
    * Guess  it's possible to add one for missing test cases.

3. also thought about that Something like

   > ok 1 Test_Param fail (run x of  y for parameter 'z')
   
    but this may produce long lines depending on the input parameters. And maybe cause issues in the formatters?

**Comment by martin-schulze-vireso** [comment]
Sorry for Bring unclear. I see the other issue as laying ground work fir impletmenting this more easily. Ince dynamic test registration is in olace, this issue is more about adding syntactic sugar than about fundamental changes. I'd like to have both and to avoid duplicating features, so my prefereed solution would be to implement dynamic test registration first and to add parametrization afterwards.

**Comment by martin-schulze-vireso** [comment]
Sorry for Bring unclear. I see the other issue as laying ground work for impletmenting this more easily. Once dynamic test registration is in place, this issue is more about adding syntactic sugar than about fundamental changes. 

I'd like to have both and to avoid duplicating features, so my preferred solution would be to implement dynamic test registration first and to add parametrization afterwards.

**Comment by irkode** [comment]
sounds like a deal - heading forward to see that coming.

**Comment by martin-schulze-vireso** [comment]
@irkode (and everybody else who is interested in this issue): I just published [v1.11.0-RC1](https://github.com/bats-core/bats-core/releases/tag/v1.11.0-RC1) and would like some feedback. Does it work for your usecase, does it break existing test suites. Thanks in advance for participating!

**Comment by edsantiago** [comment]
Thank you! Looks promising, but there seems to be a weird bug when function args have spaces?
```bash
#!/usr/bin/env bats   -*- bats -*-

function foo() {
    #    echo "# got here $*" >&3
    echo "# ${BATS_TEST_NAME}" >&3
    true
}

@test "pre test" {
    foo "pre test"
    true
}

for i in "ab" "c d"; do
    bats_test_function --description "myname $RANDOM" -- foo "$i"
done

@test "post test" {
    foo "post test"
    true
}
```
```console
$ bats test/system/506-foo.bats 
506-foo.bats
 ✓ myname 7532
   foo ab
 ✓                <<<<<<<<<<------------- missing test name
 ✓ pre test
   test_pre_test
 ✓ post test
   test_post_test

4 tests, 0 failures
```

**Comment by martin-schulze-vireso** [comment]
@edsantiago This issue should be fixed in https://github.com/bats-core/bats-core/pull/856

**Comment by irkode** [comment]
Sry. Slipped through.
We shifted toolchain et all to Windows 
Handling powershell and pester now.
So no platform to test that

**Comment by martin-schulze-vireso** [comment]
As I said previously, this issue is mainly about syntactic sugar. Your suggestion would be possible but managing the parametrization outside if the test framework is brittle.

You can already parametrize within the test code via using bats_test_function. (See above for examples)

**Comment by jaredmcphillen** [comment]
Understood. Many thanks!

**Comment by AlexSkrypnyk** [comment]
maybe this can help

https://bats-helpers.drevops.com/data-provider

**Comment by JohnRRocha** [comment]
@martin-schulze-vireso 

> As I said previously, this issue is mainly about syntactic sugar. Your suggestion would be possible but managing the parametrization outside of the test framework is brittle.
> 
> You can already parametrize within the test code via using bats_test_function. (See above for examples)

Where is this example? The only instance of `bats_test_function` I can find in this issue is your reference to it here. I've been struggling for several hours trying to get `bats_test_function` to work, and an example would be GREATLY appreciated. It would be nice if it was in the documentation too.

If you are referring to the example in #349.... that doesn't work :cry:. I attempted that with V1.12.0 and it fails with the message

```bash
$ bats example_tests.bats
1..1
not ok 1 bats-gather-tests
# ERROR: unknown option assert_equals for bats_test_function
```

Thank you.

**Comment by jakub-bochenski** [comment]
There are some issues when using the approach demonstrated in https://github.com/bats-core/bats-core/issues/241#issuecomment-1919724659

BASH special chars will cause issues  if used in  `--description`: `(`, `)`, `"`, ```, `;`
